<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Objetivo-e-Instrucciones:" data-toc-modified-id="Objetivo-e-Instrucciones:-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Objetivo e Instrucciones:</a></span><ul class="toc-item"><li><span><a href="#Objetivo" data-toc-modified-id="Objetivo-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Objetivo</a></span></li><li><span><a href="#Fecha-de-Entrega:" data-toc-modified-id="Fecha-de-Entrega:-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Fecha de Entrega:</a></span></li><li><span><a href="#Detalles-e-instrucciones-de-la-competencia:" data-toc-modified-id="Detalles-e-instrucciones-de-la-competencia:-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Detalles e instrucciones de la competencia:</a></span></li><li><span><a href="#Reporte" data-toc-modified-id="Reporte-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Reporte</a></span></li><li><span><a href="#Baseline" data-toc-modified-id="Baseline-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Baseline</a></span></li></ul></li><li><span><a href="#1.-Introducción" data-toc-modified-id="1.-Introducción-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>1. Introducción</a></span></li><li><span><a href="#2.-Representaciones" data-toc-modified-id="2.-Representaciones-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>2. Representaciones</a></span></li><li><span><a href="#3.-Algoritmos" data-toc-modified-id="3.-Algoritmos-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>3. Algoritmos</a></span></li><li><span><a href="#4.-Métricas-de-Evaluación" data-toc-modified-id="4.-Métricas-de-Evaluación-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>4. Métricas de Evaluación</a></span></li><li><span><a href="#5.-Experimentos" data-toc-modified-id="5.-Experimentos-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>5. Experimentos</a></span><ul class="toc-item"><li><span><a href="#Importar-librerías-y-utiles" data-toc-modified-id="Importar-librerías-y-utiles-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Importar librerías y utiles</a></span></li><li><span><a href="#Definir-métodos-de-evaluación" data-toc-modified-id="Definir-métodos-de-evaluación-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>Definir métodos de evaluación</a></span></li><li><span><a href="#Datos" data-toc-modified-id="Datos-6.3"><span class="toc-item-num">6.3&nbsp;&nbsp;</span>Datos</a></span></li><li><span><a href="#Analizar-los-datos" data-toc-modified-id="Analizar-los-datos-6.4"><span class="toc-item-num">6.4&nbsp;&nbsp;</span>Analizar los datos</a></span></li><li><span><a href="#Custom-Features" data-toc-modified-id="Custom-Features-6.5"><span class="toc-item-num">6.5&nbsp;&nbsp;</span>Custom Features</a></span></li><li><span><a href="#Definir-la-representación-y-el-clasificador" data-toc-modified-id="Definir-la-representación-y-el-clasificador-6.6"><span class="toc-item-num">6.6&nbsp;&nbsp;</span>Definir la representación y el clasificador</a></span></li><li><span><a href="#Ejecutar-el-pipeline-para-algún-dataset" data-toc-modified-id="Ejecutar-el-pipeline-para-algún-dataset-6.7"><span class="toc-item-num">6.7&nbsp;&nbsp;</span>Ejecutar el pipeline para algún dataset</a></span></li><li><span><a href="#Ejecutar-el-sistema-creado-por-cada-train-set" data-toc-modified-id="Ejecutar-el-sistema-creado-por-cada-train-set-6.8"><span class="toc-item-num">6.8&nbsp;&nbsp;</span>Ejecutar el sistema creado por cada train set</a></span></li><li><span><a href="#Predecir-los-target-set-y-crear-la-submission" data-toc-modified-id="Predecir-los-target-set-y-crear-la-submission-6.9"><span class="toc-item-num">6.9&nbsp;&nbsp;</span>Predecir los target set y crear la submission</a></span></li></ul></li><li><span><a href="#6.-Conclusiones" data-toc-modified-id="6.-Conclusiones-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>6. Conclusiones</a></span></li></ul></div>

# Tarea 1 NLP : Competencia de Clasificación de Texto
-------------------------------


- **Nombre:** Juan Pablo Cáceres y Javier Urrutia

- **Usuario o nombre de equipo en Codalab:** PanteraNegra




## Objetivo e Instrucciones:

### Objetivo

Esta tarea consiste en participar en una competencia cuyo objetivo es la clasificación de tweets según su intensidad de emoción. Específicamente: 

Tendrán 4 datasets de tweets de distintas emociones: `anger`, `fear`, `sadness` y `joy`. Para cada uno de estos datasets, deberán crear un clasificador que indique la intensidad de dicha emoción en sus tweets (`low`, `medium`, `high`).

###  Fecha de Entrega: 

Por ser anunciada una vez termine el paro. Se publicará la fecha en ucursos.

### Detalles e instrucciones de la competencia:

- La competencia consiste en resolver 4 problemas de clasificación distintos, cada uno de tres clases. Por cada problema deberán crear un clasificador distinto. La evaluación de la competencia se realiza en base a 4 métricas: AUC, Kappa y Accuracy. Los mejores puntajes en cada ítem serán los que ganen.

- Para comenzar se les entregará en este notebook el baseline y la estructura del reporte. El baseline es el código que realiza creación de features y clasificación básica. Los puntajes de este serán ocupados como base para la competencia: deben superar sus resultados para ser bien evaluados.  

- Para participar, deben registrarse en Codalab y luego ingresar a la competencia usando el siguiente [link]( https://competitions.codalab.org/competitions/24121?secret_key=f5eb2d95-b36e-4aad-8fc5-4d9d77f4e4dc). 

- **Es requisito entregar el reporte con el código y haber participado en la competencia para ser evaluado.**

- Pueden hacer grupos de máximo 2 alumnos. Cada grupo debe tener un nombre de equipo (En codalab, ir a settings y después cambiar Team Name). Solo una persona debe administrar la cuenta del grupo.

- En total pueden hacer un **máximo de 4 envíos/submissions** (tanto para equipos como para envíos indivuales).

- Hagan varios experimentos haciendo cross-validation o evaluación sobre una sub-partición antes de enviar sus predicciones a Codalab. Asegúrense que la distribución de las clases sea balanceada en las particiones de training y testing. Verificar que el formato de la submission coincida con el de la competencia. De lo contrario, se les será evaluado incorrectamente.

- Estar top 5 en alguna métrica equivale a 1 punto extra en la nota final.

- No se limiten a los contenidos vistos ni a scikit ni a este baseline. ¡Usen todo su conocimiento e ingenio en mejorar sus sistemas! 

- Todas las dudas escríbanlas en el hilo de U-cursos de la tarea. Los emails que lleguen al equipo docente serán remitidos a ese medio.


### Reporte

Este debe cumplir la siguiente estructura:

1.	**Introducción**: Presentar brevemente el problema a resolver, los métodos y representaciones utilizadas en el desarrollo de la tarea y conclusiones obtenidas. (0.5 Puntos)
2.	**Representaciones**: Describir los atributos y representaciones usadas como entrada de los clasificadores. Si bien, con Bag of Words (baseline) ya se comienzan a percibir buenos resultados, pueden mejorar su evaluación agregando más atributos y representaciones diseñadas a mano. Mas abajo encontrarán una lista útil de estos que les podrá ser de utilidad. (1.5 puntos)
3.	**Algoritmos**: Describir brevemente los algoritmos de clasificación usados. (0.5 puntos)
4.	**Métricas de evaluación**: Describir brevemente las métricas utilizadas en la evaluación indicando que miden y su interpretación. (0.5 puntos)
5.	**Experimentos**: Reportar todos sus experimentos. Comparar los resultados obtenidos utilizando diferentes algoritmos y representaciones. Estos experimentos los hacen sobre la sub-partición de evaluación que deben crear (o pueden usar cross-validation). Incluyan todo el código de sus experimentos aquí. ¡Es vital haber realizado varios experimentos para sacar una buena nota! (2 puntos)
6.	**Conclusiones**: Discutir resultados, proponer trabajo futuro. (1 punto)

### Baseline

Por último, el baseline contiene un código básico que:

- Obtiene los dataset.
- Divide los datasets en train (entrenamiento y prueba) y target set (el que clasificar para subir a la competencia).
- Crea un Pipeline que: 
    - Crea features personalizadas.
    - Transforma los dataset a bag of words (BoW).  
    - Entrena un clasificador usando cada train set.
- Clasifica y evalua el sistema creado usando el test set.
- Clasifica el target set.
- Genera una submission con el target en formato zip en el directorio en donde se está ejecutando el notebook. 


Algunas pistas sobre como mejorar el rendimiento de los sistemas que creen. (Esto tendrá mas sentido cuando vean el código)

- **Vectorizador**: investigar los modulos de `nltk`, en particular, `TweetTokenizer`, `mark_negation` para reemplazar los tokenizadores. También, el parámetro `ngram_range` (Ojo que el clf naive bayes no debería usarse con n-gramas, ya que rompe el supuesto de independencia). Además, implementar los atributos que crean útiles desde el listado del el enunciado. Investigar también el vectorizador tf-idf.

- **Clasificador**: investigar otros clasificadores mas efectivos que naive bayes. Estos deben poder retornar la probabilidad de pertenecia de las clases (ie: implementar la función `predict_proba`).

- **Features**: Recuerden que pueden implementar todas las features que se les ocurra! Aquí les adjuntamos algunos ejemplos:
    -	Word n-grams.
    -	Character n-grams. 
    -	Part-of-speech tags.
    -	Sentiment Lexicons (Lexicon = A set of words with a label or associated value.).
        - Count the number of positive and negative words within a sentence.
        - If the lexicon has associated intensity of feeling (for example in a decimal), then take the average of the intensity of the sentence according to the feeling, the sum, etc.
        -	A good lexicon of sentiment: [Bing Liu](http://www.cs.uic.edu/~liub/FBS/opinion-lexicon-English.rar) 
        - A reference with a lot of [sentiment lexicons](https://medium.com/@datamonsters/sentiment-analysis-tools-overview-part-1-positive-and-negative-words-databases-ae35431a470c). 
    -	The number of elongated words (words with one character repeated more than two times).
    -	The number of words with all characters in uppercase.
    -	The presence and the number of positive or negative emoticons.
    -	The number of individual negations.
    -	The number of contiguous sequences of dots, question marks and exclamation marks.
    -	Word Embeddings: Here are some good ideas on how to use them.
    https://stats.stackexchange.com/questions/221715/apply-word-embeddings-to-entire-document-to-get-a-feature-vector

- **Reducción de dimensionalidad**: También puede serles de ayuda. Referencias [aquí](https://scikit-learn.org/stable/modules/unsupervised_reduction.html).

- Por último, pueden encontrar mas referencias de cómo mejorar sus features, el vectorizador y el clasificador [aquí](https://affectivetweets.cms.waikato.ac.nz/benchmark/).

(Pueden eliminar cualquier celda con instrucciones...)

**Importante**: Recuerden poner su nombre y el de su usuario o de equipo (en caso de que aplique) tanto en el reporte. NO serán evaluados Notebooks sin nombre.

----------------------------------------

## 1. Introducción

En este trabajo se buscará diseñar modelos que permitan determinar la intensidad de la emoción que presentan distintos tweets. Para ello se utilizarán técnicas de NLP basadas en Machine Learning, extrayendo características de los textos y probando distintos tipos de clasificadores.

Los datos que se utilizararán son cuatro dataset que corresponden a tweets de cuatro emociones distintas (Enojo, miedo, alegría y tristeza) que están clasificados en tres niveles de intensidad (bajo, medio y alto).

Para resolver este problema se utilizararáán distintos tipos de representaciones como BoW, TF-IDF y Words Embeddings preentrenados. Para realizar la clasificación se probó con Naive Bayes multinomial, SVM clasification, MLP y regresión logística.


## 2. Representaciones

Se probaron distintos tipos de representaciones para los textos. 

En primer lugar están las que representan los tokens contenidos en un texto que son Bag of Words(BoW) y Term-frequency-inverse document frequency(TF-IDF). En la primera se crea un arreglo con la cuenta de los tokens presentes en el texto mientras que la segunda muestra la relevancia de cada token presente en el texto que está dada por su frecuencia en éste descontado por su frecuencia en todo el corpus de entrenamiento.

Tambien se extrae como característica la frecuencia de los carácteres '?','!' y '#' que pueden definir en parte la intensidad de lo que se está diciendo. Además de una representación de la frecuencia de los emojis que presentan los tweets.

Otra forma de representacion utilizada es el word2vec, que lleva cada palabra a un espacio vectorial que se crea al entrenar un modelo en un corpus grande para predecir palabras contexto. Así este espacio representa de cierta forma el significado de la palabra y al utilizar una funcion de agregacion, como la suma, se puede obtener un vector que represente un documento completo. Sin embargo, entrenar este tipo de modelos requiere un corpus bastante grande, por lo cual se decidio a utilizar los pesos entrenados GloVe Twitter 200d. 

Con el fin que las representaciones obtener representaciones densas más informativas, se evaluo en varias representaciones la utilización de análisis de componentes principales (PCA), lo cual realiza una selección de las dimensiones que presentan mayor información.

## 3. Algoritmos

Se probaron distintos tipos de clasificador para elegir el que tuviera mejor rendimiento.

El primero es un Naive Bayes multinomial que se utiliza directamente sobre la representación BoW y calcula la probabilidad de que un documento pertenezca a cada clase.

El segundo es un *support vector clasification*(SVC) que utiliza SVM para clasificar cada clase en un esquema uno versus uno. Este se probó von un kernel lineal y *rbf*. En la implementación de scikit se soporta además el entrenamiento sobre un set desbalanceado al ajustar automáticamente un peso para cada clase.

El tercer clasificador utilizado es una red neuronal *Multi Layer Perceptron*(MLP). Para elegir la cantintad de neuronas y la función de activación se realizó una búsqueda con la función *gridsearch* de scikit learn, sobre uno de los conjuntos.

El cuarto clasificador es un aregresión logistica que itiliza un esquema uno versus el resto para el caso multiclase. En este tambíen se realizo un ajuste de peso para cada clase por el desbalance de los datos de entrenamiento.

## 4. Métricas de Evaluación

Se utilizan las siguiente métricas para evaluar los modelos entrenados
•	AUC: Corresponde al área bajo la curva ROC(Receiver Operating Characteristics) que se obtiene al graficar el ratio de verdaderos positivos. Como este caso se trabaja con 3 clases, se calcula como el promedio ponderado del AUC de cada clase versus el resto.
•	Kappa: Corresponde a un estadístico que compara la exactitud del clasificador (Accuracy) con la exactitud esperada por un clasificador aleatorio y dependerá del balance de las clases clasificadas.
•	Accuracy: Corresponde a la exactitud del clasificador y es la proporción entre las clasificaciones correctas(verdaderos positivos) y el total de ejemplos clasificados.



## 5. Experimentos

A continuación se muestran los experimentos realizados.

En primer lugar se probaron la representaciones BoW junto con *CharsCountTransformer* con los distintos clasificadores propuestos para elegir el más adecuado. Luego se implemento la representación word embedding con GloVe Twitter 200d junto con *CharsCountTransformer* con distintos tipos tipos de clasificadores. PCA se utilizo unicamente en los experimentos que provocaba mejoras en las metricas definidas en la sección anterior. Con respecto a la representación *EmojiCountTranformer* esta no presento mejoras en ninguno de los experimentos.

Para evaluar los modelos se sigue una estrategia de validación cruzada dividiendo el set de entrenamiento con la estrategia *Stratified K-Fold* en 5 sets. Se utiliza ésta pues mantiene el balance de datos que existe originalmente en el set en todas las particiones.

### Importar librerías y utiles

In [0]:
import pandas as pd
import os
import multiprocessing
import numpy as np
import shutil
import string 

from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.pipeline import Pipeline, FeatureUnion
from sklearn.metrics import confusion_matrix, cohen_kappa_score, classification_report, accuracy_score, roc_auc_score
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.base import BaseEstimator, TransformerMixin
from imblearn.over_sampling import RandomOverSampler, SMOTE
import imblearn.pipeline

import nltk 
nltk.download('stopwords')  
from nltk.corpus import stopwords 
from sklearn.feature_extraction.text import TfidfVectorizer  
from nltk.sentiment.util import mark_negation
from nltk.tokenize import TweetTokenizer 

/usr/local/lib/python3.6/dist-packages/sklearn/externals/six.py:31: FutureWarning: The module is deprecated in version 0.21 and will be removed in version 0.23 since we've dropped support for Python 2.7. Please rely on the official version of six (https://pypi.org/project/six/).
  "(https://pypi.org/project/six/).", FutureWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/utils/deprecation.py:144: FutureWarning: The sklearn.neighbors.base module is  deprecated in version 0.22 and will be removed in version 0.24. The corresponding classes / functions should instead be imported from sklearn.neighbors. Anything that cannot be imported from sklearn.neighbors is now part of the private API.
  warnings.warn(message, FutureWarning)


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


/usr/local/lib/python3.6/dist-packages/nltk/twitter/__init__.py:20: UserWarning: The twython library has not been installed. Some functionality from the twitter package will not be available.
  warnings.warn("The twython library has not been installed. "


### Definir métodos de evaluación

Estas funciones están a cargo de evaluar los resultados de la tarea. No deberían cambiarlas.


In [0]:
def auc_score(test_set, predicted_set):
    high_predicted = np.array([prediction[2] for prediction in predicted_set])
    medium_predicted = np.array(
        [prediction[1] for prediction in predicted_set])
    low_predicted = np.array([prediction[0] for prediction in predicted_set])
    high_test = np.where(test_set == 'high', 1.0, 0.0)
    medium_test = np.where(test_set == 'medium', 1.0, 0.0)
    low_test = np.where(test_set == 'low', 1.0, 0.0)
    auc_high = roc_auc_score(high_test, high_predicted)
    auc_med = roc_auc_score(medium_test, medium_predicted)
    auc_low = roc_auc_score(low_test, low_predicted)
    auc_w = (low_test.sum() * auc_low + medium_test.sum() * auc_med +
             high_test.sum() * auc_high) / (
                 low_test.sum() + medium_test.sum() + high_test.sum())
    return auc_w


def evaulate(predicted_probabilities, y_test, labels, dataset_name):
    # Importante: al transformar los arreglos de probabilidad a clases,
    # entregar el arreglo de clases aprendido por el clasificador.
    # (que comunmente, es distinto a ['low', 'medium', 'high'])
    predicted_labels = [
        labels[np.argmax(item)] for item in predicted_probabilities
    ]
    print('Confusion Matrix for {}:\n'.format(dataset_name))
    print(
        confusion_matrix(y_test,
                         predicted_labels,
                         labels=['low', 'medium', 'high']))

    print('\nClassification Report:\n')
    print(
        classification_report(y_test,
                              predicted_labels,
                              labels=['low', 'medium', 'high']))
    # Reorder predicted probabilities array.
    labels = labels.tolist()
    predicted_probabilities = predicted_probabilities[:, [
        labels.index('low'),
        labels.index('medium'),
        labels.index('high')
    ]]
    auc = round(auc_score(y_test, predicted_probabilities), 3)
    print("Scores:\n\nAUC: ", auc, end='\t')
    kappa = round(cohen_kappa_score(y_test, predicted_labels), 3)
    print("Kappa:", kappa, end='\t')
    accuracy = round(accuracy_score(y_test, predicted_labels), 3)
    print("Accuracy:", accuracy)
    print('------------------------------------------------------\n')
    return np.array([auc, kappa, accuracy])

def score_fun(predicted_probabilities, y_test, labels):
    predicted_labels = [
        labels[np.argmax(item)] for item in predicted_probabilities
    ]
    # Reorder predicted probabilities array.
    labels = labels.tolist()
    predicted_probabilities = predicted_probabilities[:, [
        labels.index('low'),
        labels.index('medium'),
        labels.index('high')
    ]]
    auc = round(auc_score(y_test, predicted_probabilities), 3)
    kappa = round(cohen_kappa_score(y_test, predicted_labels), 3)
    accuracy = round(accuracy_score(y_test, predicted_labels), 3)
    return np.array([auc, kappa, accuracy])

### Datos

Obtener los datasets desde el github del curso

In [0]:
# Datasets de entrenamiento.
train = {
    'anger': pd.read_csv('https://raw.githubusercontent.com/dccuchile/CC6205/master/assignments/assignment_1/data/train/anger-train.txt', sep='\t', names=['id', 'tweet', 'class', 'sentiment_intensity']),
    'fear': pd.read_csv('https://raw.githubusercontent.com/dccuchile/CC6205/master/assignments/assignment_1/data/train/fear-train.txt', sep='\t', names=['id', 'tweet', 'class', 'sentiment_intensity']),
    'joy': pd.read_csv('https://raw.githubusercontent.com/dccuchile/CC6205/master/assignments/assignment_1/data/train/joy-train.txt', sep='\t', names=['id', 'tweet', 'class', 'sentiment_intensity']),
    'sadness': pd.read_csv('https://raw.githubusercontent.com/dccuchile/CC6205/master/assignments/assignment_1/data/train/sadness-train.txt', sep='\t', names=['id', 'tweet', 'class', 'sentiment_intensity'])
}
# Datasets que deberán predecir para la competencia.
target = {
    'anger': pd.read_csv('https://raw.githubusercontent.com/dccuchile/CC6205/master/assignments/assignment_1/data/target/anger-target.txt', sep='\t', names=['id', 'tweet', 'class', 'sentiment_intensity'], na_values=['NONE']),
    'fear': pd.read_csv('https://raw.githubusercontent.com/dccuchile/CC6205/master/assignments/assignment_1/data/target/fear-target.txt', sep='\t', names=['id', 'tweet', 'class', 'sentiment_intensity'], na_values=['NONE']),
    'joy': pd.read_csv('https://raw.githubusercontent.com/dccuchile/CC6205/master/assignments/assignment_1/data/target/joy-target.txt', sep='\t', names=['id', 'tweet', 'class', 'sentiment_intensity'], na_values=['NONE']),
    'sadness': pd.read_csv('https://raw.githubusercontent.com/dccuchile/CC6205/master/assignments/assignment_1/data/target/sadness-target.txt', sep='\t', names=['id', 'tweet', 'class', 'sentiment_intensity'], na_values=['NONE'])
}

In [0]:
# Ejemplo de algunas filas aleatorias:
train['anger'].sample(5)

id  ... sentiment_intensity
282  10282  ...              medium
780  10780  ...                 low
810  10810  ...                 low
119  10119  ...                high
387  10387  ...              medium

[5 rows x 4 columns]

### Analizar los datos 

Imprimir la cantidad de tweets de cada dataset, según su intensidad de sentimiento. Noten que las clases están desbalanceadas. 

In [0]:
def get_group_dist(group_name, train):
    print(group_name, "\n",
          train[group_name].groupby('sentiment_intensity').count(),
          '\n---------------------------------------\n')
for dataset_name in train:
    get_group_dist(dataset_name, train)

anger 
                       id  tweet  class
sentiment_intensity                   
high                 163    163    163
low                  161    161    161
medium               617    617    617 
---------------------------------------

fear 
                       id  tweet  class
sentiment_intensity                   
high                 270    270    270
low                  288    288    288
medium               699    699    699 
---------------------------------------

joy 
                       id  tweet  class
sentiment_intensity                   
high                 195    195    195
low                  219    219    219
medium               488    488    488 
---------------------------------------

sadness 
                       id  tweet  class
sentiment_intensity                   
high                 197    197    197
low                  210    210    210
medium               453    453    453 
---------------------------------------



### Custom Features 

Para crear features personalizadas implementaremos nuestros propios Transformers (estandar de scikit para crear nuevas features entre otras cosas). Para esto:

1. Creamos nuestra clase Transformer extendiendo BaseEstimator y TransformerMixin. En este ejemplo, definiremos `CharsCountTransformer` que cuenta carácteres relevantes ('!', '?', '#') en los tweets.
2. Definios una función cómo `get_relevant_chars` que opera por cada tweet y retorna un arreglo.
3. Hacemos un override de la función `transform` en donde iteramos por cada tweet, llamamos a la función que hicimos antes y agregamos sus resultados a un arrelo. Finalmente lo retornamos.

Esto nos facilitará el trabajo mas adelante. Una Guia completa de las transformaciones predefinidas en scikit pueden encontrarla [aquí](https://scikit-learn.org/stable/data_transforms.html).



In [0]:
class CharsCountTransformer(BaseEstimator, TransformerMixin):
    def get_relevant_chars(self, tweet):
        num_hashtags = tweet.count('#')
        num_exclamations = tweet.count('!')
        num_interrogations = tweet.count('?')
        return [num_hashtags, num_exclamations, num_interrogations]

    def transform(self, X, y=None):
        chars = []
        for tweet in X:
            chars.append(self.get_relevant_chars(tweet))

        return np.array(chars)

    def fit(self, X, y=None):
        return self

In [0]:
# Veamos que sucede si ejecutamos el transformer
sample = train['anger'].sample(5).tweet
pd.DataFrame(zip(sample, CharsCountTransformer().transform(sample)))

0          1
0  That way ur so angry you can literally feel ur...  [0, 0, 0]
1  @SapientSoldier timid but determined version o...  [0, 0, 0]
2  Controlling #anger is a sign of those #righteo...  [6, 0, 0]
3  Nahhhhhh @konanplaydirty snap story has got me...  [0, 0, 0]
4  Forever angry that gh ruined Molly and morgan'...  [0, 0, 0]

In [0]:
class EmojiCountTransformer(BaseEstimator, TransformerMixin):
    def __init__(self):
      emoticons = set(range(int('1f600',16), int('1f650', 16)))
      self.emoji = [c for c in emoticons]
      self.hot_emoji = [0 for emo in emoticons]

    def get_relevant_chars(self, tweet):
        for w in tweet:
          for char in w:
            if ord(char) in self.emoji:
              i = self.emoji.index(ord(char))
              self.hot_emoji[i]+=1
        #print(self.hot_emoji)
        return self.hot_emoji

    def transform(self, X, y=None):
        chars = []
        for tweet in X:
            chars.append(self.get_relevant_chars(tweet))

        return np.array(chars)

    def fit(self, X, y=None):
        return self

In [0]:
# Veamos que sucede si ejecutamos el transformer
sample = train['joy'].sample(5).tweet
pd.DataFrame(zip(sample, EmojiCountTransformer().transform(sample)))

0                                                  1
0  @MerenthaProphet - the Hunter in this way, con...  [0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...
1  @Nikhilv95 @LydiajaneF I wish, I really, truly...  [0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...
2  #LMFAO @MSNBC 's #racepimp Tamron Hall used th...  [0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...
3  @BertGatti not had much to cheer about you see...  [0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...
4  @elena_yip My random guesses for you! A potent...  [0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...

### Doc2vec

In [0]:
from collections import Counter
import string 
string.punctuation
from nltk.tokenize import TweetTokenizer # a tweet tokenizer from nltk.

class Doc2VecTransformer(BaseEstimator, TransformerMixin):
    """ Transforma tweets a representaciones vectoriales usando algún modelo de Word Embeddings.
    """
    def __init__(self, model, aggregation_func):
        # extraemos los embeddings desde el objeto contenedor. ojo con esta parte.
        self.model = model 
        self.tokenizer = TweetTokenizer()
        self.table = str.maketrans('', '', string.punctuation)
        # indicamos la función de agregación (np.min, np.max, np.mean, np.sum, ...)
        self.aggregation_func = aggregation_func

    def simple_tokenizer(self, doc, lower=False):
        """Tokenizador. Elimina signos de puntuación, lleva las letras a minúscula(opcional) y 
           separa el tweet por espacios.
        """  
        doc = doc.lower()
        doc = self.tokenizer.tokenize(doc)
        doc = [token for token in doc if not token.startswith('@')] # eliminar usuarios
        doc = [w.translate(self.table) for w in doc]
        doc = [w for w in doc if w != ''] # elimina words ''
        doc = [w for w in doc if w.isalpha()]
        doc = [w for w in doc if w in self.model.vocab]
        return doc

    def transform(self, X, y=None):
        
        doc_embeddings = []
        
        for doc in X:
            # tokenizamos el documento. Se llevan todos los tokens a minúscula. 
            # ojo con esto, ya que puede que tokens con minúscula y mayúscula tengan
            # distintas representaciones
            tokens = self.simple_tokenizer(doc, lower = True) 
            #print(tokens)
            selected_wv = []
            for token in tokens:
                if token in self.model.vocab:
                    selected_wv.append(self.model[token])
                else:
                  #pass
                  print(token)
                    
            # si seleccionamos por lo menos un embedding para el tweet, lo agregamos y luego lo añadimos.
            if len(selected_wv) > 0:
                doc_embedding = self.aggregation_func(np.array(selected_wv), axis=0)
                doc_embeddings.append(doc_embedding)
            # si no, añadimos un vector de ceros que represente a ese documento.
            else: 
                print('No pude encontrar ningún embedding en el tweet: {}. Agregando vector de ceros.'.format(doc))
                doc_embeddings.append(np.zeros(self.model.vector_size)) # la dimension del modelo 
        #print(np.array(doc_embeddings).shape)
        return np.array(doc_embeddings)   

    def fit(self, X, y=None):
        return self  


In [0]:
import gensim.downloader as api
word_vectors = api.load("glove-twitter-200")

[================================================--] 97.3% 737.9/758.5MB downloaded

In [0]:
print('size vocab w2v:',len(word_vectors.vocab))
word_vectors.vector_size
#word_vectors.most_similar("cant")

size vocab w2v: 1193514


200

In [0]:
# Veamos que sucede si ejecutamos el transformer
import sys
sample = train['anger'].sample(5).tweet
pd.DataFrame(zip(sample, Doc2VecTransformer(word_vectors, np.sum).transform(sample)))

0                                                  1
0  #Taurus will react angrily when she can't take...  [0.43582892, 4.9305, -0.44725996, 2.711693, -0...
1  @notquitefrodo @dominictarason I remember TAs ...  [4.7843857, 2.916362, 3.7790437, 0.924433, 1.0...
2  Petition for non-Maghrebi PoC to stop buying '...  [-0.55734116, 2.375447, 1.0665569, 4.1605077, ...
3  @evanareteos :)) im now writing abt the changi...  [0.6360725, 0.78656703, 0.073327, 0.8076659, -...
4  @SusanHensThe shameful display I watched today...  [5.535181, 7.109447, -0.40932912, 1.170087, -3...

### Negation Tweet

In [0]:
tweetTok = TweetTokenizer(reduce_len=True, strip_handles=True)

In [0]:
def negationTweetTokenizer(doc):
  global tweetTok
  return mark_negation(tweetTok.tokenize(doc))

In [0]:
for tweet in sample:
  print(negationTweetTokenizer(tweet))

['#Taurus', 'will', 'react', 'angrily', 'when', 'she', "can't", 'take_NEG', 'being_NEG', 'provoked_NEG', 'any_NEG', 'longer_NEG', '.']
['I', 'remember', 'TAs', 'in', 'the', 'gamedev', 'program', 'said', 'he', 'had', 'temper', 'tantrums', 'in', 'class', 'all', 'the', 'time', 'lol']
['Petition', 'for', 'non-Maghrebi', 'PoC', 'to', 'stop', 'buying', "'", 'couscous', "'", 'from', 'white', 'grocery', 'stores', 'and', 'boiling', 'it', '.', 'That', 'is', 'not', 'couscous_NEG', ',_NEG', 'that_NEG', 'is_NEG', 'trash_NEG', '.']
[':)', ')', 'im', 'now', 'writing', 'abt', 'the', 'changing', 'face', 'of', 'sex', 'industry', 'in', 'tr', ':)', ')', ')', 'now', 'men', 'will', 'talk', 'to', 'me', 'then', '!']
['shameful', 'display', 'I', 'watched', 'today', 'has', 'left', 'me', 'reeling', 'with', 'so', 'much', 'anger', 'dt', 'I', 'feel', 'like', 'exploding', ',', 'those', 'clowns', 'should', 'watch', 'it']


### Definir la representación y el clasificador

Para esto, definiremos Pipelines. Un `Pipeline` es una lista de transformaciones y un estimador(clasificador) ubicado al final el cual define el flujo que seguiran nuestros datos dentro del sistema que creemos. Nos permite ejecutar facilmente el mismo proceso sobre todos los datasets que usemos, simplificando así nuestra programación.

El pipeline más básico que podemos hacer es transformar el dataset a Bag of Words y después usar clasificar el BoW usando NaiveBayes:

```python
    Pipeline([('bow', CountVectorizer()), ('clf', MultinomialNB())])
```


Ahora, si queremos usar nuestra transformación para agregar las features que creamos, usaremos `FeatureUnion`. Esta simplemente concatenará los vectores resultantes de ejecutar BoW y los Transformer en un solo vector.

```python
    Pipeline([('features',FeatureUnion([('bow', CountVectorizer()),
                                        ('chars_count',CharsCountTransformer())])),
              ('clf', MultinomialNB())])

```





Recuerden que cada pipeline representa un sistema de clasificación distinto. Por lo mismo, deben instanciar uno por cada problema que resuelvan. De lo contrario, podrían solapar resultados.  Para esto, les recomendamos crear los pipeline en distintas funciones, como la siguiente:

### Features: CountVectorizer

In [0]:
def get_experiment_0_pipeline():
    return Pipeline([('features',
                      FeatureUnion([('bow', CountVectorizer()),
                                    ('chars_count', CharsCountTransformer())
                                    ])), ('clf', MultinomialNB())])

############################ SVM ##############################################

def get_experiment_search_0_pipeline():
    parameters = [{'kernel': ['rbf'], 'gamma': [1e-3, 1e-4],
                     'C': [1, 10, 100, 1000]},
                  {'kernel': ['linear'], 'C': [1, 10, 100, 1000]}]
    cores = multiprocessing.cpu_count()
    return Pipeline([('features',
                      FeatureUnion([('bow', CountVectorizer()),('chars_count', CharsCountTransformer())])), 
       ('clf', GridSearchCV(SVC(probability=True,class_weight='balanced'), parameters, n_jobs=cores,verbose=2, scoring='roc_auc_ovr_weighted'))              ])

    
def get_experiment_1_pipeline():
    return Pipeline([('features',
                      FeatureUnion([('bow', CountVectorizer()),
                                    ('chars_count', CharsCountTransformer())
                                    ])), ('clf', SVC(kernel='linear',C=1, probability=True, class_weight='balanced'))])
    
def get_experiment_2_pipeline():
    return Pipeline([('features',
                      FeatureUnion([('bow', CountVectorizer()),
                                    ('chars_count', CharsCountTransformer())
                                    ])), ('clf', SVC(kernel='rbf',C=1, probability=True, class_weight='balanced'))])
    
def get_experiment_3_pipeline():
    return Pipeline([('features',
                      FeatureUnion([('bow', CountVectorizer()),
                                    ('chars_count', CharsCountTransformer())
                                    ])), ('clf', SVC(kernel='rbf',C=1000, gamma=0.001, probability=True, class_weight='balanced'))])

######################## MLP ###################################################

def get_experiment_search_1_pipeline():
    # MLP gridsearch
    parameters = {'hidden_layer_sizes':[(100,),(200,),(100,50),(50,),(50,50)],
                  'activation':('relu', 'logistic','tanh')}
    cores = multiprocessing.cpu_count()
    return Pipeline([('features',
                      FeatureUnion([('bow', CountVectorizer()),
                                    ('chars_count', CharsCountTransformer())
                                    ])),
                     ('clf', GridSearchCV(MLPClassifier(), parameters, n_jobs=cores,verbose=2, scoring='roc_auc_ovr_weighted'))])


def get_experiment_4_pipeline():
    return imblearn.pipeline.Pipeline([('features',
                      FeatureUnion([('bow', CountVectorizer()),
                                    ('chars_count', CharsCountTransformer())
                                    ])),
                                    ('sampler', RandomOverSampler()),
                                    ('clf', MLPClassifier(hidden_layer_sizes=(100,100), activation='logistic'))])

def get_experiment_5_pipeline():
    return imblearn.pipeline.Pipeline([('features',
                      FeatureUnion([('bow', CountVectorizer()),
                                    ('chars_count', CharsCountTransformer())
                                    ])),
                                    ('sampler', RandomOverSampler()),
                                    ('clf', MLPClassifier(hidden_layer_sizes=(200,), activation='tanh', max_iter=300))])
    
######################## Logistic ###################################################    
def get_experiment_6_pipeline():
    return Pipeline([('features',
                      FeatureUnion([('bow', CountVectorizer()),
                                    ('chars_count', CharsCountTransformer())
                                    ])),
                                    ('clf', LogisticRegression(class_weight='balanced', max_iter=300))])
    


### Features: Tf-idf

In [0]:
######################## tf-idf ###################################################
def get_experiment_7_pipeline():
    return Pipeline([('features',TfidfVectorizer(max_features=2000, min_df=5, max_df=0.8, stop_words=stopwords.words('english'))  ),
                     ('clf', SVC(kernel='linear',C=1, probability=True, class_weight='balanced'))])
    
def get_experiment_8_pipeline():
    global tweetTok
    return Pipeline([('features',
                      FeatureUnion([('tf-idf', TfidfVectorizer(max_features=2000, min_df=5, max_df=0.8, tokenizer=tweetTok.tokenize)),
                                    ('chars_count', CharsCountTransformer())
                                    ])),
                     ('clf', SVC(kernel='linear',C=1, probability=True, class_weight='balanced'))])
    
def get_experiment_9_pipeline():
    global tweetTok
    return Pipeline([('features',
                      FeatureUnion([('tf-idf', TfidfVectorizer(max_features=2000, min_df=5, max_df=0.8, tokenizer=negationTweetTokenizer)),
                                    ('chars_count', CharsCountTransformer())
                                    ])),
                     ('clf', SVC(kernel='linear',C=1, probability=True, class_weight='balanced'))])

def get_experiment_10_pipeline():
    global tweetTok
    return Pipeline([('features',
                      FeatureUnion([('tf-idf', TfidfVectorizer(max_features=4000, min_df=5, max_df=0.8, tokenizer=tweetTok.tokenize)),
                                    ('chars_count', CharsCountTransformer())
                                    ])),
                     ('clf', SVC(kernel='linear',C=1, probability=True, class_weight='balanced'))])

### Features: Word2Vec

In [0]:
from sklearn.decomposition import PCA
def get_experiment_11_pipeline():
    pca = PCA()
    return Pipeline([('features',
                      FeatureUnion([('features1',Doc2VecTransformer(word_vectors, np.mean)), 
                                    ('chars_count', CharsCountTransformer())])),('pca', pca),
                                    ('clf', SVC(kernel='rbf',C=1000,gamma=0.0001, probability=True, class_weight='balanced'))])
    
def get_experiment_12_pipeline(): # Ganando
    pca = PCA()
    return Pipeline([('features',
                      FeatureUnion([('features1',Doc2VecTransformer(word_vectors, np.mean)),('emoji_count',EmojiCountTransformer()),
                                    ('chars_count', CharsCountTransformer())
                                    ])),('pca', pca),
                                    ('clf', LogisticRegression(class_weight='balanced'))])
    
def get_experiment_13_pipeline():
    pca = PCA()
    return imblearn.pipeline.Pipeline([('features',
                      FeatureUnion([('features1',Doc2VecTransformer(word_vectors, np.mean)),
                                    ('chars_count', CharsCountTransformer())
                                    ])),
                                    ('clf', MLPClassifier(hidden_layer_sizes=(50,50),solver='adam', activation='logistic'))])

def get_experiment_search_2_pipeline():
    # MLP gridsearch
    pca = PCA()
    parameters = {'hidden_layer_sizes':[(100,),(200,),(100,50),(50,),(50,50),(100,100)],
                  'activation':('relu', 'logistic','tanh'),'solver' : ['lbfgs', 'sgd', 'adam'],}
    cores = multiprocessing.cpu_count()
    return Pipeline([('features',FeatureUnion([('features1',Doc2VecTransformer(word_vectors, np.mean)),
                                    ('chars_count', CharsCountTransformer())
                                    ])),
                     ('clf', GridSearchCV(MLPClassifier(), parameters, n_jobs=cores,verbose=2, scoring='roc_auc_ovr_weighted'))])
    
def get_experiment_search_3_pipeline():
    # SVM gridsearch
    pca = PCA()
    parameters = [{'kernel': ['rbf'], 'gamma': [1e-3, 1e-4],
                     'C': [1, 10, 100, 1000]},
                  {'kernel': ['linear'], 'C': [1, 10, 100, 1000]}]
    cores = multiprocessing.cpu_count()
    return Pipeline([('features',
                      FeatureUnion([('features',Doc2VecTransformer(word_vectors, np.mean)),
                                    ('chars_count', CharsCountTransformer())
                                    ])),
                     ('clf', GridSearchCV(SVC(probability=True,class_weight='balanced'), parameters, n_jobs=cores,verbose=2, scoring='roc_auc_ovr_weighted'))])


### Ejecutar el pipeline para algún dataset

In [0]:
from sklearn.utils import class_weight
from sklearn.model_selection import StratifiedKFold
from collections import Counter

def run_cv(dataset, pipeline_generator):
    # Dividir por stratified k-fold
    skf = StratifiedKFold(n_splits=5, shuffle=False)
    scores = np.zeros((5,3))
    i=0
    X = dataset.tweet
    y = dataset.sentiment_intensity
    count_t = Counter(y)
    print(f"total:{len(dataset)} L:{count_t['low']} M:{count_t['medium']} H:{count_t['high']}")
    for train_index, test_index in skf.split(X, y):
        X_train, X_test = X[train_index], X[test_index]
        y_train, y_test = y[train_index], y[test_index]
        count_train = Counter(y_train)
        count_test = Counter(y_test)
        print("split_{}  total:{}/{}  L:{}/{}  M:{}/{}  H:{}/{}"
              .format(i, len(train_index), len(test_index),
                      count_train['low'],count_test['low'],
                      count_train['medium'],count_test['medium'],
                      count_train['high'],count_test['high']))

        # Generar pipeline
        pipeline = pipeline_generator()
        
        # Entrenamos el clasificador (Ejecuta el entrenamiento sobre todo el pipeline)
        pipeline.fit(X_train, y_train)
        
        # Predecimos las probabilidades de intensidad de cada elemento del set de prueba.
        predicted_probabilities = pipeline.predict_proba(X_test)
    
        # Obtenemos el orden de las clases aprendidas.
        learned_labels = pipeline.classes_

        # Evaluamos:
        scores[i] = score_fun(predicted_probabilities, y_test, learned_labels)
        i+=1
    return scores.mean(axis=0)

def run_eval(dataset, dataset_name, pipeline, seed=None):
    """Creamos el pipeline y luego lo ejecutamos el pipeline sobre un dataset. 
    Retorna el modelo ya entrenado mas sus labels asociadas y los scores obtenidos al evaluarlo."""

    # Dividimos el dataset en train y test.
    X_train, X_test, y_train, y_test = train_test_split(
        dataset.tweet,
        dataset.sentiment_intensity,
        shuffle=True,
        test_size=0.33,
        random_state=seed)
    
    class_weights = class_weight.compute_class_weight('balanced',
                                                 np.unique(y_train),
                                                 y_train)
    
    # Entrenamos el clasificador (Ejecuta el entrenamiento sobre todo el pipeline)
    pipeline.fit(X_train, y_train)

    # Predecimos las probabilidades de intensidad de cada elemento del set de prueba.
    predicted_probabilities = pipeline.predict_proba(X_test)

    # Obtenemos el orden de las clases aprendidas.
    learned_labels = pipeline.classes_

    # Evaluamos:
    scores = evaulate(predicted_probabilities, y_test, learned_labels, dataset_name)
    return pipeline, learned_labels, scores

def run(dataset, pipeline):
    """Creamos el pipeline y luego lo ejecutamos el pipeline sobre un dataset. 
    Retorna el modelo ya entrenado mas sus labels asociadas"""

    # Entrenamos el clasificador (Ejecuta el entrenamiento sobre todo el pipeline)
    pipeline.fit(dataset.tweet, dataset.sentiment_intensity)

    # Obtenemos el orden de las clases aprendidas.
    learned_labels = pipeline.classes_

    return pipeline, learned_labels

### SVC gridsearch

In [0]:
svc = get_experiment_search_3_pipeline()
svc,_ = run(train['sadness'], svc)

Fitting 5 folds for each of 12 candidates, totalling 60 fits


[Parallel(n_jobs=2)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=2)]: Done  37 tasks      | elapsed:   24.5s
[Parallel(n_jobs=2)]: Done  60 out of  60 | elapsed:  1.3min finished


In [0]:
print(svc[-1].best_score_)
print(svc[-1].best_params_)

0.7191189707678747
{'C': 100, 'gamma': 0.001, 'kernel': 'rbf'}


### MLP gridsearch


In [0]:
mlp = get_experiment_search_2_pipeline()
mlp,_ = run(train['sadness'], mlp)

Fitting 5 folds for each of 54 candidates, totalling 270 fits


[Parallel(n_jobs=2)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=2)]: Done  37 tasks      | elapsed:   59.7s
[Parallel(n_jobs=2)]: Done 158 tasks      | elapsed:  3.4min
[Parallel(n_jobs=2)]: Done 270 out of 270 | elapsed:  6.1min finished
/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


In [0]:
print(mlp[-1].best_score_)
print(mlp[-1].best_params_)

0.7144067310712086
{'activation': 'logistic', 'hidden_layer_sizes': (50, 50), 'solver': 'adam'}


### Evaluar el sistema creado por cada train set

Este código crea y entrena los 4 sistemas de clasificación y luego los evalua. Para los experimentos, pueden copiar este código variando el pipeline cuantas veces estimen conveniente.

In [0]:
learned_labels_array = []
scores_array = []

# Por cada nombre_dataset, dataset en train ('anger', 'fear', 'joy', 'sadness')
for dataset_name, dataset in train.items():
    print(dataset_name)
    # ejecutamos el pipeline sobre el dataset
    scores = run_cv(dataset, get_experiment_13_pipeline)

    # guardamos los scores obtenidos
    scores_array.append(scores)
    
    print("{} average scores:\n".format(dataset_name))
    print("Average AUC: {0:.3g}\t Average Kappa: {1:.3g}\t Average Accuracy: {2:.3g}\n\n"
          .format(*scores))
          
# print avg scores
print(
    "Average scores:\n\n",
    "Average AUC: {0:.3g}\t Average Kappa: {1:.3g}\t Average Accuracy: {2:.3g}"
    .format(*np.array(scores_array).mean(axis=0)))

anger
total:941 L:161 M:617 H:163
split_0  total:752/189  L:128/33  M:494/123  H:130/33


/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


split_1  total:753/188  L:129/32  M:494/123  H:130/33


/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


split_2  total:753/188  L:129/32  M:494/123  H:130/33


/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


split_3  total:753/188  L:129/32  M:493/124  H:131/32


/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


split_4  total:753/188  L:129/32  M:493/124  H:131/32


/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


anger average scores:

Average AUC: 0.705	 Average Kappa: 0.281	 Average Accuracy: 0.648


fear
total:1257 L:288 M:699 H:270
split_0  total:1005/252  L:230/58  M:559/140  H:216/54


/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


split_1  total:1005/252  L:230/58  M:559/140  H:216/54


/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


split_2  total:1006/251  L:231/57  M:559/140  H:216/54


/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


split_3  total:1006/251  L:231/57  M:559/140  H:216/54


/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


split_4  total:1006/251  L:230/58  M:560/139  H:216/54


/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


fear average scores:

Average AUC: 0.736	 Average Kappa: 0.301	 Average Accuracy: 0.608


joy
total:902 L:219 M:488 H:195
split_0  total:721/181  L:175/44  M:390/98  H:156/39


/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


split_1  total:721/181  L:175/44  M:390/98  H:156/39


/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


split_2  total:722/180  L:176/43  M:390/98  H:156/39


/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


split_3  total:722/180  L:175/44  M:391/97  H:156/39


/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


split_4  total:722/180  L:175/44  M:391/97  H:156/39


/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


joy average scores:

Average AUC: 0.734	 Average Kappa: 0.339	 Average Accuracy: 0.607


sadness
total:860 L:210 M:453 H:197
split_0  total:688/172  L:168/42  M:363/90  H:157/40


/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


split_1  total:688/172  L:168/42  M:363/90  H:157/40


/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


split_2  total:688/172  L:168/42  M:362/91  H:158/39


/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


split_3  total:688/172  L:168/42  M:362/91  H:158/39


/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


split_4  total:688/172  L:168/42  M:362/91  H:158/39
sadness average scores:

Average AUC: 0.714	 Average Kappa: 0.329	 Average Accuracy: 0.601


Average scores:

 Average AUC: 0.722	 Average Kappa: 0.313	 Average Accuracy: 0.616


/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


In [0]:
classifiers = []
learned_labels_array = []
scores_array = []

# Por cada nombre_dataset, dataset en train ('anger', 'fear', 'joy', 'sadness')
for dataset_name, dataset in train.items():
    
    # creamos el pipeline
    pipeline = get_experiment_13_pipeline()
    
    # ejecutamos el pipeline sobre el dataset
    classifier, learned_labels, scores = run_eval(dataset, dataset_name, pipeline, seed=100)

    # guardamos el clasificador entrenado (en realidad es el pipeline ya entrenado...)
    classifiers.append(classifier)

    # guardamos las labels aprendidas por el clasificador
    learned_labels_array.append(learned_labels)

    # guardamos los scores obtenidos
    scores_array.append(scores)

# print avg scores
print(
    "Average scores:\n\n",
    "Average AUC: {0:.3g}\t Average Kappa: {1:.3g}\t Average Accuracy: {2:.3g}"
    .format(*np.array(scores_array).mean(axis=0)))

/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


Confusion Matrix for anger:

[[ 16  30   1]
 [ 32 156  23]
 [  0  33  20]]

Classification Report:

              precision    recall  f1-score   support

         low       0.33      0.34      0.34        47
      medium       0.71      0.74      0.73       211
        high       0.45      0.38      0.41        53

    accuracy                           0.62       311
   macro avg       0.50      0.49      0.49       311
weighted avg       0.61      0.62      0.61       311

Scores:

AUC:  0.661	Kappa: 0.194	Accuracy: 0.617
------------------------------------------------------



/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


Confusion Matrix for fear:

[[ 41  55   5]
 [ 32 166  26]
 [  5  54  31]]

Classification Report:

              precision    recall  f1-score   support

         low       0.53      0.41      0.46       101
      medium       0.60      0.74      0.67       224
        high       0.50      0.34      0.41        90

    accuracy                           0.57       415
   macro avg       0.54      0.50      0.51       415
weighted avg       0.56      0.57      0.56       415

Scores:

AUC:  0.708	Kappa: 0.244	Accuracy: 0.573
------------------------------------------------------



/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


Confusion Matrix for joy:

[[ 35  33   1]
 [ 21 120  18]
 [  0  37  33]]

Classification Report:

              precision    recall  f1-score   support

         low       0.62      0.51      0.56        69
      medium       0.63      0.75      0.69       159
        high       0.63      0.47      0.54        70

    accuracy                           0.63       298
   macro avg       0.63      0.58      0.60       298
weighted avg       0.63      0.63      0.62       298

Scores:

AUC:  0.757	Kappa: 0.358	Accuracy: 0.631
------------------------------------------------------

Confusion Matrix for sadness:

[[ 33  25   1]
 [ 31 102  23]
 [  3  27  39]]

Classification Report:

              precision    recall  f1-score   support

         low       0.49      0.56      0.52        59
      medium       0.66      0.65      0.66       156
        high       0.62      0.57      0.59        69

    accuracy                           0.61       284
   macro avg       0.59      0.59      0.

/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


Average AUC: 0.71	 Average Kappa: 0.287	 Average Accuracy: 0.609

### Entrenar el sistema con cada train set completo

In [0]:
classifiers = []
learned_labels_array = []

# Por cada nombre_dataset, dataset en train ('anger', 'fear', 'joy', 'sadness')
for dataset_name, dataset in train.items():
    
    # creamos el pipeline
    pipeline = get_experiment_13_pipeline()
    
    # ejecutamos el pipeline sobre el dataset
    classifier, learned_labels = run(dataset, pipeline)

    # guardamos el clasificador entrenado (en realidad es el pipeline ya entrenado...)
    classifiers.append(classifier)

    # guardamos las labels aprendidas por el clasificador
    learned_labels_array.append(learned_labels)

/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


### Predecir los target set y crear la submission

Aquí predecimos los target set usando los clasificadores creados y creamos los archivos de las submissions.

In [0]:
def predict_target(dataset, classifier, labels):
    # Predecir las probabilidades de intensidad de cada elemento del target set.
    predicted = pd.DataFrame(classifier.predict_proba(dataset.tweet), columns=labels)
    # Agregar ids
    predicted['id'] = dataset.id.values
    # Reordenar las columnas
    predicted = predicted[['id', 'low', 'medium', 'high']]
    return predicted

In [0]:
predicted_target = {}

# Crear carpeta ./predictions
if (not os.path.exists('./predictions')):
    os.mkdir('./predictions')

else:
    # Eliminar predicciones anteriores:
    shutil.rmtree('./predictions')
    os.mkdir('./predictions')

# por cada target set:
for idx, key in enumerate(target):
    # Predecirlo
    predicted_target[key] = predict_target(target[key], classifiers[idx],
                                           learned_labels_array[idx])
    # Guardar predicciones en archivos separados. 
    predicted_target[key].to_csv('./predictions/{}-pred.txt'.format(key),
                                 sep='\t',
                                 header=False,
                                 index=False)

# Crear archivo zip
a = shutil.make_archive('predictions', 'zip', './predictions')

### Resultados de experimentos

> Con Validación cruzada



| Feature                                               | Clasificador                | Average AUC      | anger AUC      | fear AUC      | joy AUC      | sadness AUC      |
|-------------------------------------------------------|-----------------------------|------------------|----------------|---------------|--------------|------------------|
| bow+CharsCountTransformer                             | MultinomialNB               |            0.654 |          0.624 |         0.645 |        0.703 |            0.642 |
| bow+CharsCountTransformer                             | Linear SVC C=1              |            0.683 |          0.687 |         0.707 |        0.677 |            0.662 |
| tf-idf(max 2000)+stopwords                            | Linear SVC C=1              |            0.667 |          0.646 |         0.681 |        0.677 |            0.665 |
| tf-idf(max 2000)+tweetTokenizer+CharsCountTransformer | Linear SVC C=1              |            0.681 |          0.653 |         0.670 |         0.71 |             0.69 |
| word2vect+tweetTokenizer+CharsCountTransformer+pca    | Linear SVC C=1              |            0.711 |          0.705 |         0.714 |        0.732 |            0.695 |
| word2vect+tweetTokenizer+CharsCountTransformer+pca    | Rbf SVC C=1000 gamma=0.0001 |            0.719 |          0.704 |         0.716 |        0.737 |            0.719 |
| word2vect+tweetTokenizer+CharsCountTransformer+pca    | LogisticRegression          |            0.719 |          0.703 |         0.729 |        0.743 |              0.7 |
| word2vect+tweetTokenizer+CharsCountTransformer        | MLP                         |            0.720 |          0.702 |         0.730 |        0.731 |            0.717 |

|

| Feature | Clasificador | Average Kappa | anger Kappa | fear Kappa | joy Kappa | sadness Kappa |
|-|-|-|-|-|-|-|
| bow+CharsCountTransformer | MultinomialNB | 0.177 | 0.165 | 0.151 | 0.258 | 0.134 |
| bow+CharsCountTransformer | Linear SVC C=1 | 0.183 | 0.168 | 0.240 | 0.222 | 0.1 |
| tf-idf(max 2000)+stopwords | Linear SVC C=1 | 0.161 | 0.105 | 0.172 | 0.197 | 0.171 |
| tf-idf(max 2000)+tweetTokenizer+CharsCountTransformer | Linear SVC C=1 | 0.197 | 0.160 | 0.176 | 0.258 | 0.194 |
| word2vect+tweetTokenizer+CharsCountTransformer+pca | Linear SVC C=1 | 0.219 | 0.150 | 0.236 | 0.26 | 0.23 |
| word2vect+tweetTokenizer+CharsCountTransformer+pca | Rbf SVC C=1000 gamma=0.0001 | 0.242 | 0.183 | 0.262 | 0.267 | 0.255 |
| word2vect+tweetTokenizer+CharsCountTransformer+pca | LogisticRegression | 0.302 | 0.301 | 0.303 | 0.338 | 0.268 |
| word2vect+tweetTokenizer+CharsCountTransformer | MLP | 0.315 | 0.291 | 0.285 | 0.336 | 0.348 |

|

| Feature | Clasificador | Average Accuracy | anger Accuracy | fear Accuracy | joy Accuracy | sadness Accuracy |
|-|-|-|-|-|-|-|
| bow+CharsCountTransformer | MultinomialNB | 0.600 | 0.666 | 0.572 | 0.606 | 0.554 |
| bow+CharsCountTransformer | Linear SVC C=1 | 0.603 | 0.672 | 0.611 | 0.593 | 0.538 |
| tf-idf(max 2000)+stopwords | Linear SVC C=1 | 0.589 | 0.664 | 0.572 | 0.567 | 0.55 |
| tf-idf(max 2000)+tweetTokenizer+CharsCountTransformer | Linear SVC C=1 | 0.612 | 0.683 | 0.588 | 0.602 | 0.576 |
| word2vect+tweetTokenizer+CharsCountTransformer+pca | Linear SVC C=1 | 0.618 | 0.669 | 0.608 | 0.607 | 0.589 |
| word2vect+tweetTokenizer+CharsCountTransformer+pca | Rbf SVC C=1000 gamma=0.0001 | 0.623 | 0.675 | 0.616 | 0.604 | 0.599 |
| word2vect+tweetTokenizer+CharsCountTransformer+pca | LogisticRegression | 0.558 | 0.577 | 0.551 | 0.574 | 0.532 |
| word2vect+tweetTokenizer+CharsCountTransformer | MLP | 0.618 | 0.65 | 0.602 | 0.605 | 0.614 |

## 6. Conclusiones

Observando los resultados de los modelos entrenados con Cross Validation, se observar que los mejores modelos corresponden al feature: *word2vect+tweetTokenizer+CharCountsTransformer+PCA*, con clasificadores SVM con kernel RBF y MLP. Al evaluar estos modelos cor la función *eval* se obtiene que el mejor clasificador corresponde a MLP, con métricas de Average AUC: 0.71	 Average Kappa: 0.287	 Average Accuracy: 0.609. 

Que el mejor modelo corresponda a MLP, se debe a que las características con la cual este algortimo discrimina corresponden a vectores densos, los cuales son ideales para MLP. 

Se pudo observar que en general representaciones con word embeddings, entregan mejor accuracy que representaciones basadas en conteo, las cuales pueden contener varios 0s. Cabe mencionar, que el conteo de carácteres corresponde a una buena representación que permite agregar buena información a las características bases probadas. Por otro lado, el conteo de emojis implementado no fue una técnica para entregar información, debido a que no aporto a mejorar las métricas, ya que era un vector demasiado sparse. 

Finalmente, se menciona que la actividad pudo ser realizada de forma exitosa, pudiendo experimentar con varios métodos de extracción de caractererísticas y algoritmos de clasificación, permitiendo corroborar la teoría. 